---
title: "Dilution series analysis"
author: "Vincent Caruso"
date: "August 25, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Set up the environment.
```{r setup}

library(dada2)
library(stringr)
library(Biostrings)
library(tidyverse)
library(ShortRead)
library(RColorBrewer)

options(tibble.print_min = 40, tibble.print_max = 40)
# options(tibble.width = Inf)

data_path <- "~/data/dilution_separate"
result_path <- "~/results/dilution_separate"
merged_path <- file.path(result_path, "merged")
ref_path <- "~/references"
analysis_path <- "~/analysis/dilution_separate"
script_path <- "~/noisy-microbes/community-inference/scripts"

source(file.path(script_path, "seq_table_processing.R"))

if (!file_test("-d", file.path(analysis_path, "blast"))) dir.create(file.path(analysis_path, "blast"), recursive = TRUE)
if (!file_test("-d", file.path(analysis_path, "summary"))) dir.create(file.path(analysis_path, "summary"), recursive = TRUE)
if (!file_test("-d", merged_path)) dir.create(merged_path)

```

##Load and format OTU/SV tables from each method

Load and inspect the tables from each clustering method. Modify tables as necessary so that they all have the following format:

1. Rows are OTUs or SVs
2. The first column is the OTU/SV identifier, and the remaining columns are samples' abundances
3. There are no comment lines (just a header with column names)

After modifying the format of each table, write it back to a tab-separated '.txt' file.
```{r load inference results}

methods <- c("uclust", "uparse","med", "unoise", "deblur", "dada2")
samples <- str_subset(list.files(result_path), "^s\\d{3}")

for (m in methods){
  if (!file_test("-d", file.path(merged_path, m))) dir.create(file.path(merged_path, m))
}

```


```{r load UCLUST}

# load UCLUST table
uclust_otu_file <- "uclust/otu_table.txt"
uclust_seqs_file <- "uclust/rep_set/pooled_nochim_rep_set.fasta"
uclust_table <- list()

for (s in samples){
  uclust_table[[s]] <- load_uclust(file.path(result_path, s, uclust_otu_file), file.path(result_path, s, uclust_seqs_file))
}
names(uclust_table) <- str_replace_all(names(uclust_table), '-', '\\.')

uclust_table <- merge_tables(uclust_table, sample_names = names(uclust_table), collapse = FALSE, id = "keep")
sample_names <- colnames(uclust_table)[c(-1, -length(colnames(uclust_table)))]
uclust_table <- remove_rare(uclust_table, sample_names, min_abund = 2)  # remove singletons from all samples
uclust_table
write_tsv(uclust_table %>% select(-sequence), file.path(merged_path, "uclust/uclust_table.txt"))

```


```{r load UPARSE}

# load UPARSE table
uparse_otu_file <- "uparse/otu_table.txt"
uparse_seqs_file <- "uparse/otus.fa"
uparse_table <- list()

for (s in samples){
  uparse_table[[s]] <- load_uparse(file.path(result_path, s, uparse_otu_file),
                                   file.path(result_path, s, uparse_seqs_file))
}
names(uparse_table) <- str_replace_all(names(uparse_table), '-', '\\.')

uparse_table <- merge_tables(uparse_table, sample_names = names(uparse_table), collapse = FALSE, id = "keep")
uparse_table <- remove_rare(uparse_table, sample_names, min_abund = 2)  # remove singletons
uparse_table
write_tsv(uparse_table %>% select(-sequence), file.path(merged_path, "uparse/uparse_table.txt"))

```


```{r load MED}

# load MED table
med_otu_file <- "med/MATRIX-COUNT.txt"
med_seqs_file <- "med/NODE-REPRESENTATIVES.fasta"
med_chimera_file <- "med/uchime/NODE-chimeras.fasta"
med_table <- list()

for (s in samples){
  med_table[[s]] <- load_med(file.path(result_path, s, med_otu_file), 
                             file.path(result_path, s, med_seqs_file), 
                             file.path(result_path, s, med_chimera_file))
}
names(med_table) <- str_replace_all(names(med_table), "-", "\\.")

med_table <- merge_tables(med_table, sample_names = names(med_table), collapse = FALSE, id = "keep")
med_table <- remove_rare(med_table, sample_names, min_abund = 2)  # remove singletons
med_table
write_tsv(med_table %>% select(-sequence), file.path(merged_path, "med/med_table.txt"))

```


```{r load UNOISE}

# load UNOISE table
unoise_otu_file <- "unoise/zotu_table.txt"
unoise_seqs_file <- "unoise/zotus.fa"
unoise_table <- list()

for (s in samples){
  unoise_table[[s]] <- load_unoise(file.path(result_path, s, unoise_otu_file),
                                   file.path(result_path, s, unoise_seqs_file))
}
names(unoise_table) <- str_replace_all(names(unoise_table), "-", "\\.")

unoise_table <- merge_tables(unoise_table, sample_names = names(unoise_table), collapse = FALSE, id = "keep")
unoise_table <- remove_rare(unoise_table, sample_names, min_abund = 2)  # remove singletons
unoise_table
write_tsv(unoise_table %>% select(-sequence), file.path(merged_path, "unoise/unoise_table.txt"))

```


```{r load Deblur table}

# load Deblur table
deblur_otu_file <- "deblur/all.txt"
deblur_table <- list()

for (s in samples){
  deblur_table[[s]] <- load_deblur(file.path(result_path, s, deblur_otu_file))
}
names(deblur_table) <- str_replace_all(names(deblur_table), "-", "\\.")

deblur_table <- merge_tables(deblur_table, sample_names = names(deblur_table), collapse = FALSE, id = "keep")

deblur_table <- remove_rare(deblur_table, sample_names, min_abund = 2)  # remove singletons
deblur_table
write_tsv(deblur_table %>% select(-sequence), file.path(merged_path, "deblur/deblur_table.txt"))

```


```{r load DADA2 table}

# load DADA2 table
dada2_otu_file <- file.path(result_path, "dada2/sv_table.no_chim.txt")
dada2_table <- load_dada2(dada2_otu_file)
dada2_table <- remove_rare(dada2_table, sample_names, min_abund = 2)  # remove singletons
dada2_table
write_tsv(dada2_table %>% select(-sequence), file.path(merged_path, "dada2/dada2_table.txt"))

```


##Create and annotate a list of method tables

Create a list of method tables.
```{r create method table list}

method_tables <- list(uclust_table, uparse_table, med_table, unoise_table, deblur_table, dada2_table)
names(method_tables) <- methods

```


Read in reference sequences, and compute Levenshtein distances of inferred sequences to the reference sequences for each method.
```{r compute distance to references}

# ref_fastas <- list.files(zymo_ref, pattern = "_16S.fasta$")
ref_fasta <- "Zymo_16S_references_v2.fasta"
ref_seqs <- readDNAStringSet(file.path(ref_path, ref_fasta)) %>% as.character()
names(ref_seqs) <- str_replace(names(ref_seqs$zymo), "_16S.*", "")

ref_dist_tables <- lapply(method_tables, compute_ref_dist, ref_seqs)
strain_dist_tables <- lapply(ref_dist_tables, collapse_group_dist, names(ref_seqs))
```


Annotate sequences in sequence tables as "reference" if they exactly match a reference sequence, or as "ref_noisy" if they are within a specified Levenshtein distance range from a reference sequence. 
```{r annotate reference}

max_dist = 10
method_tables <- mapply(annotate_ref, method_tables, ref_dist_tables, 
                        MoreArgs = list(sample_names = sample_names, max_dist = max_dist),
                        SIMPLIFY = FALSE)

```


Next, write each method's inferred (singleton-free) sequences to a fasta file, for BLASTing against nt.
```{r write fastas}

mapply(write_fasta, method_tables, file.path(analysis_path, "blast", paste(methods, "seqs.fasta", sep = "_")))

# BLAST each method's sequences against the 'nt' database using the following command:
# blastn -query method_seqs.fasta -db nt -out method_blast.txt \
#        -task megablast -max_target_seqs 10 \
#        -outfmt "7 qseqid qlen sseqid slen sskingdoms ssciname pident length nident mismatch gapopen gaps qstart qend sstart send evalue bitscore" \
#        -remote

```


Read in the results of the BLAST searches.
```{r load blast tables}

blast_files <- file.path(analysis_path, "blast", paste(methods, "blast.txt", sep = "_"))
blast_tables <- lapply(blast_files, load_blast)
names(blast_tables) <- methods
blast_tables <- lapply(blast_tables, annotate_blast_table)

```


Annotate sequences with distance to the nearest NT sequence, and label them as 'Contaminant' if they exactly match an NT sequence, using the results of the BLAST search.
```{r annotate contaminant}

method_tables <- mapply(annotate_nt_dist, method_tables, blast_tables, SIMPLIFY = FALSE)

method_tables <- mapply(annotate_contam, method_tables, blast_tables, 
                        MoreArgs = list(sample_names = sample_names, max_dist = 10), 
                        SIMPLIFY = FALSE)

```


Compute distance from sequences to Contaminant sequences.
```{r compute distance to contams}

contam_seqs <- lapply(method_tables, function(mt) return(mt %>% filter(contaminant) %>% .[["sequence"]]))
contam_dist_tables <- mapply(compute_ref_dist, method_tables, contam_seqs, SIMPLIFY = FALSE)

```


Annotate Contam_Noisy, Other, and Consensus, then add a 'Class' factor column with the sequence classification of each sequence, with the five main classification levels.
```{r annotate contam noisy}

method_tables <- mapply(annotate_contam_like, method_tables, contam_dist_tables, blast_tables,
                        MoreArgs = list(sample_names = sample_names, noisy_dist = 10, mut_dist = 1),
                        SIMPLIFY = FALSE)

method_tables <- lapply(method_tables, annotate_class)

```


Compute distances between `contam_noisy` and `contaminant` sequences. This should give a sense of how many `contam_noisy` sequences may have been generated through sequencer errors from `contaminant` template sequences. Also, compute distances between `contaminant` sequences. This will give an idea of whether or not OTU methods underestimate the number of `contaminant` sequences by lumping them together, if they are within the 97% similarity threshold of each other.
```{r contam_noisy distances}


noisy_to_contam <- lapply(method_tables, compute_inter_dist, class1 = "contam_noisy", class2 = "contaminant")
method_tables <- mapply(annotate_inter_dist, method_tables, noisy_to_contam, 
                        MoreArgs = list(column_name = "dist_to_contam"), SIMPLIFY = FALSE)

# table minimum distances of contam_noisy sequences to contaminant sequences
for (m in methods){
  cat(m, "distances of contam_noisy to contaminant sequences:\n")
  print(noisy_to_contam[[m]] %>% apply(., 1, min) %>% table())
}


contam_to_contam <- lapply(method_tables, compute_inter_dist, class1 = "contaminant", class2 = "contaminant")
method_tables <- mapply(annotate_inter_dist, method_tables, contam_to_contam, 
                        MoreArgs = list(column_name = "dist_to_contam"), SIMPLIFY = FALSE)

# table minimum distances of contam_noisy sequences to contaminant sequences
for (m in methods){
  cat(m, "distances of contam_noisy to contaminant sequences:\n")
  print(contam_to_contam[[m]] %>% apply(., 1, function(d) min(d[d > 0])) %>% table())
}

```


##Create a list of tables that summarize method results

Compute summary tables for each method, with total sequence counts, sequence counts in each class, and percentages of reads in each class.
```{r compute method summaries}

method_summary <- mapply(summarize_seqs, method_tables, ref_dist_tables, 
                         MoreArgs = list(refs = ref_seqs, sample_names = sample_names, max_dist = max_dist),
                         SIMPLIFY = FALSE)
# names(method_summary) <- c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2")
sample_summary <- transpose_table_list(method_summary, "sample", "method")
sample_summary_short <- lapply(sample_summary, function(ss){
  return(ss %>% select(method, total, reference, ref_noisy, contaminant, contam_noisy, other))
})

method_summary <- lapply(method_summary, sanity_check_summary)
sample_summary <- lapply(sample_summary, sanity_check_summary)

lapply(method_summary, function(ms) sum(ms$total != ms$check_sum))
lapply(sample_summary, function(ss) sum(ss$total != ss$check_sum))

write_tables(method_summary, file.path(analysis_path, "method_summary_tables.tsv"))
write_tables(sample_summary, file.path(analysis_path, "sample_summary_tables.tsv"))
write_tables(sample_summary_short, file.path(analysis_path, "sample_summary_short_tables.tsv"))

```


From the tables, compute recall and precision for each method, using both sequences and reads as the unit of analysis.
```{r method summary stats}

method_seq_stats <- lapply(method_summary, compute_pr_seqs)
sample_seq_stats <- lapply(sample_summary, compute_pr_seqs)

method_read_stats <- lapply(method_tables, compute_pr_reads, sample_names)
sample_read_stats <- transpose_table_list(method_read_stats, "sample", "method")

method_ref_perc <- lapply(method_tables, compute_ref_perc, sample_names = sample_names)
method_ref_perc <- do.call("rbind", method_ref_perc) %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as.tibble()

write_tables(method_seq_stats, file.path(analysis_path, "method_seq_stats.tsv"))
write_tables(sample_seq_stats, file.path(analysis_path, "sample_seq_stats.tsv"))
write_tables(method_read_stats, file.path(analysis_path, "method_read_stats.tsv"))
write_tables(sample_read_stats, file.path(analysis_path, "sample_read_stats.tsv"))
write_tsv(method_ref_perc, file.path(analysis_path, "method_ref_perc.tsv"))

```


##Create various plots to visualize results

Add a column that assigns a color code to each sequence
```{r add color column}

library(RColorBrewer)

ref_colors <- brewer.pal(8, "Dark2") %>% rev()
ref_colors <- c(ref_colors, brewer.pal(3, "Set1"))
max_contams <- max(sapply(method_tables, function(mt) nrow(mt[!mt$reference,])))
other_colors <- rep("#A1A1A1", times = max(sapply(method_tables, function(dt) nrow(dt[!dt$reference,]))))

method_tables <- lapply(method_tables, function(mt){
  mt$color <- character(length = nrow(mt))
  mt$color[mt$reference] <- ref_colors[1:nrow(mt[mt$reference,])]
  mt$color[!mt$reference] <- other_colors[1:nrow(mt[!mt$reference,])]
  return(mt)
})

```


Reformat annotated sequence tables for plotting with `ggplot`.
```{r prep for ggplot}

# Combine all method tables into one large table, then gather the sample columns into a single column
method_tables_gg <- lapply(method_tables, gather_samples, sample_names, sample)
method_tables_gg <- lapply(method_tables_gg, annotate_norms, group = "sample")
all_table_gg <- do.call("rbind", method_tables_gg) %>% mutate(method = rep(methods, times = sapply(method_tables_gg, nrow)))
all_table_gg$method <- factor(all_table_gg$method, levels = c("uclust", "uparse", "med", "unoise", "deblur", "dada2"),
                              labels = c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2"))

# add some grouping variables for plotting convenience
all_table_gg <- all_table_gg %>% mutate(algorithm = method)
levels(all_table_gg$algorithm) <- list(OTU = c("UCLUST", "UPARSE"), entropy = "MED", EM = c("UNOISE", "Deblur", "DADA2"))

# Create plotting labels
dilution_labels <- sapply(str_split(sample_names, "\\."), function(d){
  paste(d[3:4], collapse = ":")
})
dilution_labels[1] <- "1:1"
dilution_values <- sapply(str_split(dilution_labels, ":"), function(d) as.numeric(d[1]) / as.numeric(d[2]))


```



Plot composition of samples in terms of sequence classification, facetted by method.
```{r class sequences composition}

library(ggpubr)

class_colors <- brewer.pal(12, "Paired")[c(2, 1, 8, 7, 10)]
classes = levels(all_table_gg$class)

class_seqs_comp <- ggplot(all_table_gg %>% filter(count > 0), aes(x = sample)) +
  geom_bar(aes(fill = fct_rev(class)), width = 0.7) +
  facet_wrap(~method) +
  scale_fill_manual(name = "ASV/OTU class   ", values = rev(class_colors), labels = paste(rev(classes), "  ")) +
  scale_x_discrete(labels = dilution_labels) +
  # labs(title = "Composition of dilution series communities by sequence class", subtitle = "", 
  labs(x = "", y = "number of ASVs/OTUs") +
  big_labels(title = 72, text = 60, legend.position = "top") +
  # big_labels(title = 16, text = 12, legend.position = "top") +
  theme(panel.spacing = unit(3, "lines"), 
        legend.margin = margin(b = 90)) +
  guides(fill = guide_legend(reverse = TRUE, keywidth = 4.5, keyheight = 4.5))
class_seqs_comp


class_reads_comp <- ggplot(all_table_gg %>% filter(count > 0), aes(x = sample)) +
  geom_col(aes(y = count, fill = fct_rev(class)), position = position_fill(), width = 0.7) +
  scale_x_discrete(labels = dilution_labels) +
  scale_fill_manual(name = "ASV/OTU class   ", values = rev(class_colors), labels = paste(rev(classes), "  ")) +
  facet_wrap(~method, nrow = 2) +
  # labs(title = "Read-level composition by sequence class across dilution series", subtitle = "", 
  labs(x = "sample dilution", y = "fraction of sample") +
  big_labels(title = 72, text = 60, legend.position = "top") +
  # big_labels(title = 16, text = 12, legend.position = "top") +
  theme(panel.spacing = unit(3, "lines"),
        legend.margin = margin(b = 90)) +
  guides(fill = FALSE)
  # guides(fill = guide_legend(reverse = TRUE, keywidth = 4.5, keyheight = 4.5))
class_reads_comp

ggarrange(class_seqs_comp, class_reads_comp, labels = c("A", "B"), vjust = c(1.5, 0.5), nrow = 2, 
          font.label = list(size = 96), heights = c(5, 4))


```


Plot counts of Contaminant, Contam_Noisy, and Other, faceted by method, to show whether these categories exhibit the same trend or not.
```{r non-reflike and reflike sequence lines}


classes = levels(all_table_gg$class)
class_colors <- brewer.pal(8, "Accent")[c(8, 7, 1, 6, 5)]
class_colors <- brewer.pal(9, "Set1")[c(9, 5, 4, 1, 2)]
class_colors <- brewer.pal(12, "Paired")[c(2, 1, 8, 7, 10)]
ref_class_colors <- brewer.pal(12, "Paired")[c(2, 1)] #, brewer.pal(8, "Dark2")[2])
contam_colors <- brewer.pal(12, "Paired")[c(8, 7, 10)]


reflike_seq_lines <- dilution_line_plot(all_table_gg, "class", ref_like, "count", size = 4) +
  scale_color_manual(name = "ASV/OTU class   ", values = ref_class_colors, labels = paste(classes[1:2], "   ")) +
  labs(x = "", y = "number of ASVs/OTUs") +
  facet_wrap(~method) +
  big_labels(title = 72, text = 60, legend.position = "top") +
  theme(panel.spacing = unit(3, "lines"),
        legend.margin = margin(b = 90),
        plot.margin = unit(c(2, 0, 4, 0), "lines")) +
  guides(color = guide_legend(keywidth = 8))
reflike_seq_lines

non_reflike_seq_lines <- dilution_line_plot(all_table_gg, "class", !ref_like, "count", size = 4) +
  scale_color_manual(name = "ASV/OTU class   ", values = contam_colors, labels = paste(classes[3:5], "   ")) +
  labs(x = "sample dilution", y = "number of ASVs/OTUs") +
  facet_wrap(~method) +
  big_labels(title = 72, text = 60, legend.position = "top") +
  theme(panel.spacing = unit(3, "lines"),
        legend.margin = margin(b = 90),
        plot.margin = unit(c(4, 0, 2, 0), "lines")) +
  guides(color = guide_legend(keywidth = 8))
non_reflike_seq_lines

ggarrange(reflike_seq_lines, non_reflike_seq_lines, labels = c("A", "B"), nrow = 2, font.label = list(size = 96))

```


Create plots to show relative strength of signal (reference sequence matches) to noise (non-reference sequences).
```{r snr plots}

library(scales)

snr_box_facet <- ggplot(data = all_table_gg %>% filter(count > 0)) +
  geom_boxplot(aes(x = sample, y = count, fill = factor(reference, levels = c(TRUE, FALSE))),
               outlier.size = 3, outlier.stroke = 1, size = 1.5) +
  # labs(title = "Abundance distributions of Reference and non-Reference sequences", subtitle = "", 
  labs(x = "dilution", y = expression(bold("log"[10]*" abundance"))) +
  scale_fill_manual(name = "Sequence class   ", 
                    labels = paste(c(" Reference", " non-Reference"), "   "), 
                    values = c("blue", "orange")) +
  scale_x_discrete(labels = dilution_labels) +
  scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
        labels = trans_format("log10", math_format(10^.x))) +
  facet_wrap(~method, nrow = 3, dir = "v") +
  big_labels(title = 72, text = 60, legend.position = "top") +
  theme(panel.spacing = unit(3, "lines"),
        legend.margin = margin(b = 90)) +
  guides(fill = guide_legend(keywidth = 6, keyheight = 9))
snr_box_facet

```


Create plot to show sample composition, colored by reference vs. non-reference.
```{r composition plot}

mname = "UCLUST"

set.seed(444)

comp_bar_facet <- ggplot(all_table_gg, aes(x = sample, y = count)) +
  geom_col(aes(fill = color, group = reference), width = 0.6, position = position_fill()) +
  scale_fill_identity(name = "ASV/OTU", guide = "legend",
                      labels = c(paste("Ref", 1:9), rep("non-Ref", length(unique(other_colors)))),
                      breaks = c(ref_colors[1:9], unique(other_colors))) +
  scale_x_discrete(labels = dilution_labels) +
  facet_wrap(~method) +
  labs(x = "sample dilution", y = "fraction of sample") +
  theme(legend.position = "right", 
        legend.title = element_text(face = "bold", size = 12), 
        axis.text = element_text(size = 12),
        axis.title = element_text(face = "bold", size = 16),
        plot.title = element_text(size = 20),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        strip.text = element_text(size = 16),
        panel.spacing = unit(1, "lines")) +
  guides(fill = guide_legend(reverse = FALSE, keywidth = 1, keyheight = 1))
comp_bar_facet

```


Plot alpha diversity.
```{r alpha diversity}

library("phyloseq")

# extract sequence matrices of read counts only from the method sequence tibbles
seq_mats <- lapply(method_tables, function(mt){
  seq_mat <- as.matrix(mt[, sample_names])
  rownames(seq_mat) <- mt$id
  return(seq_mat)
})


# construct phyloseq OTU and taxonomy tables
otu_tabs <- lapply(seq_mats, otu_table, taxa_are_rows = TRUE)

# construct phyloseq objects
physeqs <- lapply(otu_tabs, phyloseq)

# compute alpha diversity
alpha_indices <- c("Chao1", "Shannon", "Simpson", "InvSimpson", "Fisher")
alpha_div <- lapply(physeqs, estimate_richness, measures = alpha_indices)

# combine alpha diversity tables into one table for plotting
alpha_div <- lapply(alpha_div, rownames_to_column, var = "sample")
alpha_div <- lapply(alpha_div, as.tibble)
alpha_div_gg <- lapply(alpha_div, gather, key = "alpha_index", value = "diversity", alpha_indices)
alpha_div_gg <- do.call("rbind", alpha_div_gg) %>% 
  mutate(method = rep(methods, each = length(sample_names) * length(alpha_indices)))

alpha_colors <- brewer.pal(12, "Paired")[c(2, 3, 9, 7, 6, 12)]

alpha_dots <- ggplot(alpha_div_gg %>% filter(alpha_index %in% c("Shannon", "Simpson", "InvSimpson", "Fisher")), 
                     aes(x = sample)) +
  geom_point(aes(y = diversity, color = method), size = 2.5) +
  scale_color_manual(values = alpha_colors) +
  scale_x_discrete(labels = dilution_labels) +
  facet_wrap(~alpha_index, nrow = 1, scales = "free_y") +
  big_labels(legend.position = "top")
alpha_dots


```


Plot number of inferred sequences for each sample and each method.
```{r sequence count plots}

# normalized by median read count after processing by each method
seqs_line_norm <- ggplot(sample_counts_gg, aes(x = sample, y = seqs_norm_method_median)) +
  geom_point(aes(color = method), size = 2) +
  geom_line(aes(group = method, color = method)) +
  labs(title = "Number of inferred sequences at each concentration by each method", 
       subtitle = "Normalized by median sample read count",
       x = "sample dilution", y = "Inferred sequences (normalized)") +
  scale_color_brewer(name = "Method", palette = "Dark2") +
  scale_x_discrete(labels = dilution_labels) +
  theme(legend.position = c(0.1, 0.6),
        # legend.justification = c(0.5, 0.5),
        legend.title = element_text(face = "bold", size = 16),
        legend.text = element_text(size = 16),
        legend.background = element_rect(color = "black", size = 0.25, linetype = "solid"),
        legend.key.height = unit(20, "points"),
        axis.title = element_text(face = "bold", size = 16),
        axis.text = element_text(size = 14),
        plot.title = element_text(hjust = 0.5, size = 20),
        plot.subtitle = element_text(hjust = 0.5, size = 16))
seqs_line_norm

```
