---
title: "Dilution series analysis"
author: "Vincent Caruso"
date: "August 25, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Set up the environment.
```{r setup}

library(dada2)
library(stringr)
library(Biostrings)
library(tidyverse)

result_path <- "~/projects/thesis/results/dilution_w_blank"
ref_path <- "~/projects/thesis/references"
analysis_path <- "~/projects/thesis/analysis/dilution_w_blank"

if (!file_test("-d", analysis_path)) dir.create(analysis_path)

```

##Load OTU/SV tables from each method

Load and inspect the tables from each clustering method. Modify tables as necessary so that they all have the following format:

1. Rows are OTUs or SVs, and columns are samples
2. Row names are the OTU/SV identifiers
3. Column names are the sample names
4. There are no comment lines (just a header with column names)

After modifying the format of each table, write it back to a tab-separated '.txt' file.
```{r load UCLUST}

# load UCLUST table
uclust_table <- as.tibble(read.table(file = file.path(result_path, "uclust/otu_table.txt"), header = TRUE, sep = "\t", 
                                     skip = 1, comment.char = ""))
colnames(uclust_table)[1] <- "id"
uclust_table$id <- str_replace(uclust_table$id, "denovo", "denovo_")
sample_names <- colnames(uclust_table)[-1]
sample_names <- sample_names[order(sample_names)]
#row.names(uclust_table) <- str_replace(uclust_table[, 1], "denovo", "denovo_")  # name rows by OTU ID
#uclust_table <- uclust_table[, -1]  # remove OTU ID column
uclust_table <- uclust_table %>% select(id, noquote(sample_names))
#uclust_table <- uclust_table[, order(colnames(uclust_table))]
uclust_table
#str(uclust_table)
#uclust_table[,] <- sapply(uclust_table[,], as.integer)
write.table(uclust_table, file = file.path(result_path, "uclust/uclust_table.txt"), quote = FALSE, sep = "\t")

#dilutions <- str_subset(sample_names, "MC")

# load UCLUST representative sequences
uclust_seqs <- readDNAStringSet(file.path(result_path, "uclust/rep_set/pooled_nochim_rep_set.fasta"))
uclust_seqs <- as.character(uclust_seqs)
names(uclust_seqs) <- sapply(str_split(names(uclust_seqs), " "), `[`, 1)
names(uclust_seqs) <- str_replace(names(uclust_seqs), "denovo", "denovo_")

```


```{r load UPARSE}

# load UPARSE table
uparse_table <- as.tibble(read.table(file = file.path(result_path, "uparse/otu_table.txt"), header = TRUE, sep = "\t", 
                                     comment.char = ""))
colnames(uparse_table)[1] <- "id"
uparse_table$id <- str_replace(uparse_table$id, "OTU", "OTU_")
#uparse_table[, -1] <- uparse_table %>% select(-id) %>% .[order(colnames(uparse_table)[-1])]
colnames(uparse_table)[-1] <- sample_names
#row.names(uparse_table) <- str_replace(uparse_table[, 1], "OTU", "OTU_")  # name rows by OTU ID
#uparse_table <- uparse_table[, -1]  # remove OTU ID column
#uparse_table <- uparse_table[, order(colnames(uparse_table))]
#colnames(uparse_table) <- sample_names
uparse_table
#str(uparse_table)
write.table(uparse_table, file = file.path(result_path, "uparse/uparse_table.txt"), quote = FALSE, sep = "\t")

# load UPARSE representative sequences
uparse_seqs <- readDNAStringSet(file.path(result_path, "uparse/otus.fa"))
uparse_seqs <- as.character(uparse_seqs)
names(uparse_seqs) <- str_replace(names(uparse_seqs), "OTU", "OTU_")

```


```{r load UNOISE}

# load UNOISE table
unoise_table <- as.tibble(read.table(file = file.path(result_path, "unoise/zotu_table.txt"), header = TRUE, sep = "\t", 
                                     comment.char = ""))
colnames(unoise_table)[1] <- "id"
unoise_table$id <- str_replace(unoise_table$id, "OTU", "ZOTU_")
#row.names(unoise_table) <- str_replace(unoise_table[, 1], "OTU", "ZOTU_")  # name rows by ZOTU ID after renaming 'OTU' to 'ZOTU'
#unoise_table <- unoise_table[, -1]  # remove ZOTU ID column
#unoise_table <- unoise_table[, order(colnames(unoise_table))]
colnames(unoise_table)[-1] <- sample_names
unoise_table
#str(unoise_table)
write.table(unoise_table, file = file.path(result_path, "unoise/unoise_table.txt"), quote = FALSE, sep = "\t")

# load UNOISE sequences
unoise_seqs <- readDNAStringSet(file.path(result_path, "unoise/zotus.fa"))
unoise_seqs <- as.character(unoise_seqs)
names(unoise_seqs) <- sapply(names(unoise_seqs), FUN = str_replace, "OTU", "ZOTU_")

```


```{r load MED}

# load MED table
med_table <- read.table(file = file.path(result_path, "med/MATRIX-COUNT.txt"), header = TRUE, sep = "\t")
row.names(med_table) <- med_table[, 1]
med_table <- med_table[, -1]  # remove sample name column
med_table <- data.frame(t(med_table))   # samples as columns
med_table <- as.tibble(rownames_to_column(med_table, var = "id"))
med_table$id <- str_replace(med_table$id, "X", "Node_")
#med_table <- med_table[, order(colnames(med_table))]
med_table
#str(med_table)

# load MED sequences
med_seqs <- readDNAStringSet(file.path(result_path, "med/NODE-REPRESENTATIVES.fasta"))
med_seqs <- as.character(med_seqs)
med_seqs <- sapply(med_seqs, FUN = str_replace, "-+", "")
names(med_seqs) <- paste0("Node_", sapply(str_split(names(med_seqs), "\\|"), FUN = `[`, 1))  # remove size annotation

# remove chimeras from MED sequences
# DADA2 method
# med_seqtab <- t(med_table)
# colnames(med_seqtab) <- med_seqs
# med_seqtab.no_chim <- removeBimeraDenovo(med_seqtab)
# med_table.no_bim <- med_table[med_seqs %in% colnames(med_seqtab.no_chim), ]

# UCHIME method
med_seqs.chimeras <- readDNAStringSet(file.path(result_path, "med/uchime/NODE-chimeras.fasta"))
if (length(med_seqs.chimeras) > 0){
  med_seqs.chimeras <- as.character(med_seqs.chimeras)
  names(med_seqs.chimeras) <- paste0("Node_", sapply(str_split(names(med_seqs.chimeras), ";"), FUN = `[`, 1))
  med_table <- med_table[!med_seqs %in% med_seqs.chimeras, ]
  med_table
  med_seqs <- med_seqs[!med_seqs %in% med_seqs.chimeras]
}

write.table(med_table, file = file.path(result_path, "med/med_table.txt"), quote = FALSE, sep = "\t")

```


```{r load Deblur table}

# load Deblur table
deblur_table <- as.tibble(read.table(file = file.path(result_path, "deblur/all.txt"), header = TRUE, sep = "\t", 
                                     skip = 1, comment.char = ""))
colnames(deblur_table)[1] <- "id"
#head(deblur_table)
deblur_seqs <- toupper(deblur_table$id)  
names(deblur_seqs) <- paste0("sOTU_", 1:length(deblur_seqs))
deblur_table$id <- names(deblur_seqs)
#row.names(deblur_table) <- names(deblur_seqs)   # rename table rows with sOTU IDs intead of sequences
#deblur_table <- deblur_table[, -1]  # remove sequence column
#deblur_table <- deblur_table[, order(colnames(deblur_table))]
deblur_table <- deblur_table %>% select(id, noquote(sample_names))
deblur_table
#str(deblur_table)
#deblur_table[,] <- sapply(deblur_table[,], as.integer)
write.table(deblur_table, file = file.path(result_path, "deblur/deblur_table.txt"), quote = FALSE, sep = "\t")

```


```{r load DADA2 table}

# load DADA2 table
dada2_table <- read.table(file = file.path(result_path, "dada2/sv_table.no_chim.txt"), header = TRUE, sep = "\t")
head(unname(dada2_table))
dada2_table <- data.frame(t(dada2_table))
dada2_table <- as.tibble(rownames_to_column(dada2_table, var = "id"))
#dada2_seqs <- row.names(dada2_table)
dada2_seqs <- dada2_table$id
names(dada2_seqs) <- paste0("ASV_", 1:length(dada2_seqs))
#row.names(dada2_table) <- names(dada2_seqs)   # rename table rows with ASV IDs instead of sequences
dada2_table$id <- names(dada2_seqs)
dada2_table
colnames(dada2_table)[-1] <- sample_names
#str(dada2_table)
write.table(dada2_table, file = file.path(result_path, "dada2/dada2_table.txt"), quote = FALSE, sep = "\t")

```


Remove singletons.
```{r remove singletons}

# remove rows with no more than a single observation in any sample
uclust_table <- uclust_table[apply(uclust_table[, sample_names], 1, max) > 1, ]
uparse_table <- uparse_table[apply(uparse_table[, sample_names], 1, max) > 1, ]
unoise_table <- unoise_table[apply(unoise_table[, sample_names], 1, max) > 1, ]
med_table <- med_table[apply(med_table[, sample_names], 1, max) > 1, ]
deblur_table <- deblur_table[apply(deblur_table[, sample_names], 1, max) > 1, ]
dada2_table <- dada2_table[apply(dada2_table[, sample_names], 1, max) > 1, ]

```

Add sequences to OTU tables
```{r add sequences}

uclust_table$sequence <- uclust_seqs[uclust_table$id]
uparse_table$sequence <- uparse_seqs[uparse_table$id]
unoise_table$sequence <- unoise_seqs[unoise_table$id]
med_table$sequence <- med_seqs[med_table$id]
deblur_table$sequence <- deblur_seqs[deblur_table$id]
dada2_table$sequence <- dada2_seqs[dada2_table$id]

# remove singleton sequences from sequence vectors
uclust_seqs <- uclust_seqs[uclust_table$id]
uparse_seqs <- uparse_seqs[uparse_table$id]
unoise_seqs <- unoise_seqs[unoise_table$id]
med_seqs <- med_seqs[row.names(med_table)]
deblur_seqs <- deblur_seqs[deblur_table$id]
dada2_seqs <- dada2_seqs[dada2_table$id]

```

Read in reference sequences, and compute Levenshtein ("Hamming") distances between all inferred sequences and all reference sequences.
```{r Levenshtein function}

ref_path <- "~/projects/thesis/references"
zymo_path <- file.path(ref_path, "zymo_SSU")
ref_fastas <- list.files(zymo_path, pattern = "_16S.fasta$")

ref_seqs <- DNAStringSet()
for (f in ref_fastas){
  ref_seqs <- append(ref_seqs, readDNAStringSet(file.path(zymo_path, f)))
}
ref_seqs <- as.character(ref_seqs)
species <- sapply(str_split(names(ref_seqs), "_16S"), `[`, 1)

# define a function to compute Levenshtein distances between two sequences
levDist <- Vectorize(function(query, ref, ...) {
  mmi <- dada2:::nweval(query, ref, ...)    # this gives matches, mismatches, and indels
  ldist <- mmi[2] + mmi[3]
  if (nchar(query) > sum(mmi)) {    # query not fully overlapped by reference
    ldist <- nchar(query) - mmi[1]  # include non-overlapped nucleotides in distance
  }
  return(ldist)
})

```


Compute distances of inferred sequences to the reference sequences for each method.
```{r compute distances}

uclust_to_ref <- outer(uclust_seqs, ref_seqs, levDist, band = -1)
uparse_to_ref <- outer(uparse_seqs, ref_seqs, levDist, band = -1)
unoise_to_ref <- outer(unoise_seqs, ref_seqs, levDist, band = -1)
med_to_ref <- outer(med_seqs, ref_seqs, levDist, band = -1)
deblur_to_ref <- outer(deblur_seqs, ref_seqs, levDist, band = -1)
dada2_to_ref <- outer(dada2_seqs, ref_seqs, levDist, band = -1)

```


Write a function to count and identify sequences that differ by only 1 (or 2 or 3) necleotides from a more abundant reference sequence.
```{r noisy functions}

# function to generate a table of "noisy"" sequences: those that are very similar to the reference sequences
countNoisySeqs <- function(seq_table, dist_mat, min_dist = 1, max_dist = 3) {
  if (max(colSums(dist_mat == 0)) > 1) {
    print("WARNING: Your distance matrix contains seqeunces that are identical except for length differences.")
    print("This may result in inaccurate counts of noisy seuqences.")
  }
  
  if (sum(!sapply(seq_table, class) %in% c("integer", "numeric")) > 0){
    stop("The sequence table must only contain column vectors of type 'integer'.")
  }
  
  rmat <- matrix(0, nrow = ncol(seq_table), ncol = (max_dist - min_dist) + 1, 
                 dimnames = list(colnames(seq_table), paste(min_dist:max_dist, "off")))
  
  for (n in min_dist:max_dist){
    mins <- apply(dist_mat, 1, min)   
    noisy_ind <- which(dist_mat == n, arr.ind = TRUE)  
    noisy_ind <- noisy_ind[dist_mat[noisy_ind] == mins[noisy_ind[, "row"]], ]   # make sure it's a minimum distance
    
    if (length(noisy_ind) > 0){
      ref_row <- apply(as.matrix(dist_mat[, noisy_ind[, "col"] ]) == 0, 2, which)  # get nearest sequence that matches reference
      noisy_ind <- cbind(noisy_ind, ref_row)
      is_n_off <- ( (seq_table[noisy_ind[, "row"], ] > 0) & 
                   (seq_table[noisy_ind[, "row"], ] < seq_table[noisy_ind[, "ref_row"], ]) ) * 1  # is abundance less than reference?
      row.names(is_n_off) <- row.names(noisy_ind)
      is_n_off <- aggregate(is_n_off, by = list(row.names(is_n_off)), FUN = max)
      row.names(is_n_off) <- is_n_off$Group.1
      is_n_off <- is_n_off[, -1]
      rmat[, n] <- colSums(is_n_off)
    }
    else rmat[, n] <- integer(nrow(rmat))
  }
  rmat <- cbind(rmat, "all noisy" =  rowSums(rmat))
  return(rmat)
}

# define a function that returns whether sequences of a sequence table are "noisy"
isNoisy <- function(seq_tab, dist_mat, min_dist = 1, max_dist = 3){
  mins <- apply(dist_mat, 1, min)   # get minimum distance for each sequence
  noisy_ind <- which(dist_mat >= min_dist & dist_mat <= max_dist, arr.ind = TRUE) 
  if (length(noisy_ind) == 0){
    return(rep(FALSE, nrow(dist_mat)))
  } 
  else {
    noisy_ind <- noisy_ind[dist_mat[noisy_ind] == mins[noisy_ind[, "row"]], ]   # make sure it's a minimum distance
    ref_row <- apply(as.matrix(dist_mat[, noisy_ind[, "col"] ]) == 0, 2, which)   # find the nearest sequence that matches the reference
    noisy_ind <- cbind(noisy_ind, ref = ref_row)
    is_noisy <- ( (seq_tab[noisy_ind[, "row"], ] > 0) & 
                    (seq_tab[noisy_ind[, "row"], ] < seq_tab[noisy_ind[, "ref"], ]) )   # is abundance less than reference?
    is_noisy <- apply(is_noisy, 1, any)   # collapse to vector
    noisy_ind <- noisy_ind[is_noisy, ]    # remove any with abundance higher than reference
    noisy <- 1:nrow(dist_mat) %in% noisy_ind[, "row"]  
    return(noisy)
  }
}

```


Annotate sequences in sequence tables as "noisy" or not.
```{r annotate noisy}

uclust_table$dist_to_ref <- apply(uclust_to_ref, 1, min)
uparse_table$dist_to_ref <- apply(uparse_to_ref, 1, min)
unoise_table$dist_to_ref <- apply(unoise_to_ref, 1, min)
med_table$dist_to_ref <- apply(med_to_ref, 1, min)
deblur_table$dist_to_ref <- apply(deblur_to_ref, 1, min)
dada2_table$dist_to_ref <- apply(dada2_to_ref, 1, min)

uclust_table$hit <- uclust_table$dist_to_ref == 0
uparse_table$hit <- uparse_table$dist_to_ref == 0
unoise_table$hit <- unoise_table$dist_to_ref == 0
med_table$hit <- med_table$dist_to_ref == 0
deblur_table$hit <- deblur_table$dist_to_ref == 0
dada2_table$hit <- dada2_table$dist_to_ref == 0

uclust_table$noisy <- isNoisy(uclust_table, uclust_to_ref, 1, 3)
uparse_table$noisy <- isNoisy(uparse_table, uparse_to_ref, 1, 3)
unoise_table$noisy <- isNoisy(unoise_table, unoise_to_ref, 1, 3)
med_table$noisy <- isNoisy(med_table, med_to_ref, 1, 3)
deblur_table$noisy <- isNoisy(deblur_table, deblur_to_ref, 1, 3)
dada2_table$noisy <- isNoisy(dada2_table, dada2_to_ref, 1, 3)

uclust_table$ref_like <- uclust_table$hit | uclust_table$noisy
uparse_table$ref_like <- uparse_table$hit | uparse_table$noisy
unoise_table$ref_like <- unoise_table$hit | unoise_table$noisy
med_table$ref_like <- med_table$hit | med_table$noisy
deblur_table$ref_like <- deblur_table$hit | deblur_table$noisy
dada2_table$ref_like <- dada2_table$hit | dada2_table$noisy

```

BLAST remaining sample sequences against nt.
```{r blast high sample}

high_table$id <- paste0("seq_", 1:nrow(high_table))
high_seqs <- high_table$sequence
names(high_seqs) <- high_table$id
high_seqs <- DNAStringSet(high_seqs)
writeFasta(high_seqs, file.path(analysis_path, "high_seqs.fasta"))

blast_table <- read.table(file.path(analysis_path, "high_blast.txt"), sep = "\t", comment.char = "#",
                          col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", "gapopens", "gaps",
                                        "qstart", "qend", "sstart", "send", "evalue", "bitscore"))

```


Next, write each method's inferred (singleton-free) sequences to a fasta file, for BLASTing against nt.
```{r write fastas}

library(ShortRead)

uclust_seqs <- DNAStringSet(uclust_seqs)
writeFasta(uclust_seqs, file.path(analysis_path, "uclust_seqs.fasta"))
uparse_seqs <- DNAStringSet(uparse_seqs)
writeFasta(uparse_seqs, file.path(analysis_path, "uparse_seqs.fasta"))
unoise_seqs <- DNAStringSet(unoise_seqs)
writeFasta(unoise_seqs, file.path(analysis_path, "unoise_seqs.fasta"))
med_seqs <- DNAStringSet(med_seqs)
writeFasta(med_seqs, file.path(analysis_path, "med_seqs.fasta"))
deblur_seqs <- DNAStringSet(deblur_seqs)
writeFasta(deblur_seqs, file.path(analysis_path, "deblur_seqs.fasta"))
dada2_seqs <- DNAStringSet(dada2_seqs)
writeFasta(dada2_seqs, file.path(analysis_path, "dada2_seqs.fasta"))

```


Read in the results of the BLAST searches.
```{r load blast tables}

uclust_blast <- as.tibble(read.table(file.path(analysis_path, "uclust_blast.txt"), sep = "\t", comment.char = "#",
                                     col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
uparse_blast <- as.tibble(read.table(file.path(analysis_path, "uparse_blast.txt"), sep = "\t", comment.char = "#",
                                     col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
unoise_blast <- as.tibble(read.table(file.path(analysis_path, "unoise_blast.txt"), sep = "\t", comment.char = "#",
                                     col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
med_blast <- as.tibble(read.table(file.path(analysis_path, "med_blast.txt"), sep = "\t", comment.char = "#",
                                  col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
deblur_blast <- as.tibble(read.table(file.path(analysis_path, "deblur_blast.txt"), sep = "\t", comment.char = "#",
                                     col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
dada2_blast <- as.tibble(read.table(file.path(analysis_path, "dada2_blast.txt"), sep = "\t", comment.char = "#",
                                    col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))

```

Define functions to identify blast hits that are exact matches and those that are one or more nucleotides off.
```{r blast parsing functions}

isPerfectHit <- function(seq_tab, blast_file){
  blast_tab <- as.tibble(read.table(blast_file, sep = "\t", comment.char = "#",
                                    col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
  perfect_hits <- blast_tab[ (blast_tab$identity == 100) & 
                            (blast_tab$aln_length == nchar(seq_tab$sequence[match(blast_tab$seqID, seq_tab$id)])), ]
  return(seq_tab$id %in% perfect_hits$seqID)
}

isNOff <- function(seq_tab, blast_file, N){
  blast_tab <- as.tibble(read.table(blast_file, sep = "\t", comment.char = "#",
                                    col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
  # find full-length hits with one mismatch or indel
  fulls <- blast_tab[ blast_tab$aln_length == nchar(seq_tab$sequence[match(blast_tab$seqID, seq_tab$id)]), ]
  one_offs <- fulls[ (fulls$mismatches == 1 & fulls$gaps == 0) | (fulls$mismatches == 0 & fulls$gaps == 1), ]   #alternate code
  
  # now check length-1 hits with 100% identity
  partials <- blast_tab[ blast_tab$aln_length == (nchar(seq_tab$sequence[match(blast_tab$seqID, seq_tab$id)]) - N), ]
  one_offs <- rbind(one_offs, partials[partials$identity == 100, ])
  is_one_off <- seq_tab$id %in% one_offs$seqID
  
  hits <- isPerfectHit(seq_tab, blast_file)
  is_one_off[hits] <- FALSE   # remove queries that also have perfect hits
  
  return(is_one_off)
}

```


Annotate sequence tables with results from BLAST analysis.
```{r annotate sequence tables}

uclust_table$perfect <- isPerfectHit(uclust_table, file.path(analysis_path, "uclust_blast.txt")) & !uclust_table$hit
uclust_table$one_off <- isNOff(uclust_table, file.path(analysis_path, "uclust_blast.txt"), 1)
uclust_table <- uclust_table %>% mutate(contaminant = perfect | one_off)

uparse_table$perfect <- isPerfectHit(uparse_table, file.path(analysis_path, "uparse_blast.txt")) & !uparse_table$hit
uparse_table$one_off <- isNOff(uparse_table, file.path(analysis_path, "uparse_blast.txt"), 1)
uparse_table <- uparse_table %>% mutate(contaminant = perfect | one_off)

unoise_table$perfect <- isPerfectHit(unoise_table, file.path(analysis_path, "unoise_blast.txt")) & !unoise_table$hit
unoise_table$one_off <- isNOff(unoise_table, file.path(analysis_path, "unoise_blast.txt"), 1)
unoise_table <- unoise_table %>% mutate(contaminant = perfect | one_off)

med_table$perfect <- isPerfectHit(med_table, file.path(analysis_path, "med_blast.txt")) & !med_table$hit
med_table$one_off <- isNOff(med_table, file.path(analysis_path, "med_blast.txt"), 1)
med_table <- med_table %>% mutate(contaminant = perfect | one_off)

deblur_table$perfect <- isPerfectHit(deblur_table, file.path(analysis_path, "deblur_blast.txt")) & !deblur_table$hit
deblur_table$one_off <- isNOff(deblur_table, file.path(analysis_path, "deblur_blast.txt"), 1)
deblur_table <- deblur_table %>% mutate(contaminant = perfect | one_off)

dada2_table$perfect <- isPerfectHit(dada2_table, file.path(analysis_path, "dada2_blast.txt")) & !dada2_table$hit
dada2_table$one_off <- isNOff(dada2_table, file.path(analysis_path, "dada2_blast.txt"), 1)
dada2_table <- dada2_table %>% mutate(contaminant = perfect | one_off)

```


Now, compute summary tables by sample (for all 9 dilutions).
```{r compute sample tables}

# combine sequences from all methods
methods <- c("uclust", "uparse", "unoise", "med", "deblur", "dada2")
all_seqs <- unique(c(uclust_table$sequence, uparse_table$sequence, unoise_table$sequence, med_table$sequence, 
                     deblur_table$sequence, dada2_table$sequence))

dilution_table <- list()
dilution_to_ref <- list()
dilution_summary <- list()

for (i in seq_along(sample_names)){
  # create a sequence table for the sample
  sample_st <- matrix(as.integer(0), ncol = length(all_seqs), nrow = length(methods))
  colnames(sample_st) <- all_seqs
  rownames(sample_st) <- methods
  
  # populate the table with sample abundances
  sample_st["uclust", uclust_table$sequence] <- uclust_table[[ sample_names[i] ]]
  sample_st["uparse", uparse_table$sequence] <- uparse_table[[ sample_names[i] ]]
  sample_st["unoise", unoise_table$sequence] <- unoise_table[[ sample_names[i] ]]
  sample_st["med", med_table$sequence] <- med_table[[ sample_names[i] ]]
  sample_st["deblur", deblur_table$sequence] <- deblur_table[[ sample_names[i] ]]
  sample_st["dada2", dada2_table$sequence] <- dada2_table[[ sample_names[i] ]]

  # remove columns (sequences) with all zeros
  sample_st <- sample_st[, colSums(sample_st) > 0]
  
  # collapse sequences that only differ in length
  sample_st <- collapseNoMismatch(sample_st)
  
  # Compute distances from high-biomass sequences to reference
  sample_to_ref <- outer(colnames(sample_st), ref_seqs, levDist, band = -1)
  row.names(sample_to_ref) <- colnames(sample_st)
  # Collapse distances for strains of the same species (using minimum distance)
  sample_to_ref_species <- t(aggregate(t(sample_to_ref), list(species), min))
  colnames(sample_to_ref_species) <- sample_to_ref_species[1, ]
  sample_to_ref_species <- sample_to_ref_species[-1, ]
  class(sample_to_ref_species) <- "integer"
  
  # Convert to data.frame and annotate sequences
  sample_table <- data.frame(t(sample_st)) %>% rownames_to_column(var = "seqeunce") %>% as.tibble()
  sample_table$dist_to_ref <- apply(sample_to_ref, 1, min)
  sample_table$hit <- sample_table$dist_to_ref == 0
  sample_table$noisy <- isNoisy(sample_table, sample_to_ref)
  
  # Compute summary statistics
  sample_hits <- colSums(sample_table$hit & sample_table[, methods] > 0)
  sample_species <- integer()
  for (m in methods){
    m_species <- sum(apply(sample_to_ref_species[sample_table[[m]] > 0, ], 2, min) == 0)
    sample_species <- c(sample_species, m_species)
  }
  sample_pct_hit <- colSums(sample_table[sample_table$hit, methods]) / colSums(sample_table[, methods])
  #sample_table[, methods] <- sapply(sample_table[, methods], as.integer)
  sample_noisy <- countNoisySeqs(sample_table[, methods], sample_to_ref, 1, 3)   # "noisy" sequences
  sample_other <- colSums(sample_table$dist_to_ref > 3 & sample_table[, methods] > 0)   # non-hits, not "noisy"
  sample_counts <- colSums(sample_table[, methods] > 0)   # total number of sequences
  sample_summary <- data.frame(method = methods, total = sample_counts, hits = sample_hits, species = sample_species,
                             sample_noisy, other = sample_other, hit_rate = sample_pct_hit,
                             row.names = NULL, check.names = FALSE)
  
  dilution_table[[i]] <- sample_table
  dilution_to_ref[[i]] <- sample_to_ref
  dilution_summary[[i]] <- sample_summary
  cat("Done with sample", sample_names[i], "\n")
}

names(dilution_table) <- sample_names
names(dilution_to_ref) <- sample_names
names(dilution_summary) <- sample_names

```


Reformat annotated sequence tables for plotting with `ggplot`.
```{r prep for ggplot}

uclust_gg <- gather(uclust_table, "sample", "count", one_of(sample_names))
median_uclust <- median(colSums(uclust_table[, sample_names]))
uclust_gg <- uclust_gg %>% mutate(log10 = log10(count))
uclust_gg <- uclust_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_uclust), 
         log10_norm = log10(norm_median))
uclust_gg <- uclust_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

uparse_gg <- gather(uparse_table, "sample", "count", one_of(sample_names))
median_uparse <- median(colSums(uparse_table[, sample_names]))
uparse_gg <- uparse_gg %>% mutate(log10 = log10(count))
uparse_gg <- uparse_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_uparse), 
         log10_norm = log10(norm_median))
uparse_gg <- uparse_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

unoise_gg <- gather(unoise_table, "sample", "count", one_of(sample_names))
median_unoise <- median(colSums(unoise_table[, sample_names]))
unoise_gg <- unoise_gg %>% mutate(log10 = log10(count))
unoise_gg <- unoise_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_unoise), 
         log10_norm = log10(norm_median))
unoise_gg <- unoise_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

med_gg <- gather(med_table, "sample", "count", one_of(sample_names))
median_med <- median(colSums(med_table[, sample_names]))
med_gg <- med_gg %>% mutate(log10 = log10(count))
med_gg <- med_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_med), 
         log10_norm = log10(norm_median))
med_gg <- med_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

deblur_gg <- gather(deblur_table, "sample", "count", one_of(sample_names))
median_deblur <- median(colSums(deblur_table[, sample_names]))
deblur_gg <- deblur_gg %>% mutate(log10 = log10(count))
deblur_gg <- deblur_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_deblur), 
         log10_norm = log10(norm_median))
deblur_gg <- deblur_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

dada2_gg <- gather(dada2_table, "sample", "count", one_of(sample_names))
median_dada2 <- median(colSums(dada2_table[, sample_names]))
dada2_gg <- dada2_gg %>% mutate(log10 = log10(count))
dada2_gg <- dada2_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_dada2), 
         log10_norm = log10(norm_median))
dada2_gg <- dada2_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

```

Create tables of sequence and read count summaries per sample, and for each method.
```{r summary tables}

method_tables <- list(uclust_table, uparse_table, unoise_table, med_table, deblur_table, dada2_table)
names(method_tables) <- methods

load(file.path(result_path, "dada2/tables.RData"))
median_raw <- median(track_table[, "raw"])

sample_seq_count <- sapply(method_tables, function(st, ss) colSums(st[, ss] > 0), sample_names) %>%
  t() %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as.tibble()
sample_seq_count_gg <- gather(sample_seq_count, "sample", "seqs", -method)

sample_read_count <- sapply(method_tables, function(st, ss) colSums(st[, ss]), sample_names) %>%
  t() %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as.tibble()
sample_read_count_gg <- gather(sample_read_count, "sample", "reads", -method)

sample_summary <- merge(sample_read_count_gg, sample_seq_count_gg, by = c("method", "sample"))
sample_summary <- sample_summary %>% group_by(method) %>% 
  mutate(seqs_norm_method = round(seqs * median(reads) / reads),
         seqs_norm_raw = round(seqs * median_raw / reads))

```

Create plots to show relative strength of signal (reference hits) to noise (non-reference hits).
```{r snr plots}

library(RColorBrewer)

theme_set(theme_bw())
# ref vs. non-ref boxplots, raw counts
snr_box <- ggplot(data = deblur_gg) +
  geom_boxplot(aes(x = sample, y = log10_norm, fill = factor(ref_like, levels = c(TRUE, FALSE)))) +
  labs(title = "Distribution of log10(read counts) for reference and non-reference sequences",
          subtitle = "Dilution series processed with DADA2", x = "sample dilution", y = "log10(read counts)") +
  scale_fill_manual(name = "sequence identity", labels = c("reference", "non-reference"), values = c("blue", "orange")) +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
snr_box

# ref vs. non-ref violin plots, raw counts
snr_violin <- ggplot(data = dada2_gg) +
  geom_violin(aes(x = sample, y = log10_norm, fill = factor(hit, levels = c(T, F)))) +
  labs(title = "Density distribution of log10(read counts) for reference and non-reference sequences",
          subtitle = "Dilution series processed with DADA2", x = "sample dilution", y = "log10(read counts)") +
  scale_fill_manual(name = "sequence identity", labels = c("reference", "non-reference"), values = c("purple", "green")) +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
snr_violin

```

Create plots to show sample composition, colored by reference vs. non-reference.
```{r}

comp_bar <- ggplot(dada2_gg, aes(x = sample, y = count, fill = id, group = hit, alpha = hit)) +
  geom_col(width = 0.5, position = position_fill()) +
  #scale_fill_grey() +
  #scale_fill_manual(values = mypalette(nrow(dada2_gg))) +
  scale_fill_discrete(h = c(0, 360) + 45, c = 100, l = 45, h.start = 0, direction = 1, na.value = "grey50") +
  scale_alpha_discrete(range = c(0.25, 1)) +
  #scale_fill_identity() +
  theme(legend.position = "None")
comp_bar

```

Plot counts of inferred sequences for each sample and each method.
```{r}

# non-normalized counts from each method
seqs_line_non <- ggplot(sample_summary, aes(x = sample, y = seqs)) +
  geom_point(aes(color = method), size = 2) +
  geom_line(aes(group = method, color = method)) +
  labs(title = "Number of inferred sequences/OTUs at each concentration by each method", subtitle = "Non-normalized counts",
       x = "sample dilution", y = "sequence/OTU count") +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
#  facet_wrap(~method)
seqs_line_non

# normalized by median read count after processing by each method
seqs_line_norm <- ggplot(sample_summary, aes(x = sample, y = seqs_norm_method)) +
  geom_point(aes(color = method), size = 2) +
  geom_line(aes(group = method, color = method)) +
  labs(title = "Number of inferred sequences/OTUs at each concentration by each method", subtitle = "Normalized by median read counts",
       x = "sample dilution", y = "sequence/OTU count") +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
#  facet_wrap(~method)
seqs_line_norm

# same plot as above, but only the first seven samples
seqs_line_norm2 <- ggplot(sample_summary[sample_summary$sample %in% sample_names[1:7], ], aes(x = sample, y = seqs_norm_method)) +
  geom_point(aes(color = method)) +
  geom_line(aes(group = method, color = method)) #+
  #facet_wrap(~method)
seqs_line_norm2

# normalized by median count of raw (un-processed) reads
seqs_line_norm_raw <- ggplot(sample_summary, aes(x = sample, y = seqs_norm_raw)) +
  geom_point(aes(color = method), size = 2) +
  geom_line(aes(group = method, color = method)) +
  labs(title = "Number of inferred sequences/OTUs at each concentration by each method", subtitle = "Normalized by median read counts",
       x = "sample dilution", y = "sequence/OTU count") +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
#  facet_wrap(~method)
seqs_line_norm_raw

```


Plot distribution boxplots of non-hit read counts for "contaminant" sequences vs non-contaminant ("other") sequences.
```{r snr contaminant vs other}

snr_box_cont <- ggplot(dada2_gg[!dada2_gg$hit & !dada2_gg$noisy, ], aes(x = sample, y = log10_norm)) +
  geom_boxplot(aes(fill = factor(contaminant, levels = c(TRUE, FALSE)))) +
  #ylim(c(0, 150)) +
  labs(title = "Distribution of log10(read counts) for non-hit contaminant and non-contaminant sequences",
          subtitle = "Dilution series processed with DADA2", x = "sample dilution", y = "log10(read counts)") +
  scale_fill_manual(name = "sequence identity", labels = c("contaminant", "non-contaminant"), values = c("blue", "orange")) +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
snr_box_cont

```

