---
title: "High biomass analysis"
author: "Vincent Caruso"
date: "March 1, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Set up the environment, and define key variables.
```{r setup}

library(dada2)
library(stringr)
library(Biostrings)
library(tidyverse)
library(ShortRead)
library(RColorBrewer)

options(tibble.print_min = 100, tibble.print_max = 40)

data_path <- "~/thesis/data"
result_path <- "~/thesis/results/high_biomass"
ref_path <- "~/thesis/references"
analysis_path <- "~/thesis/analysis/high_biomass"
script_path <- "~/thesis/noisy-microbes/community-inference/scripts"
plot_path <- "~/thesis/analysis/plots"

source(file.path(script_path, "seq_table_processing.R"))

if (!file_test("-d", analysis_path)) dir.create(analysis_path)

datasets <- c("zymo", "kozich", "schirmer", "damore")
methods <- c("uclust", "uparse", "med", "unoise", "deblur", "dada2")

```


##Load OTU/SV tables from each method

Load and inspect the tables from each clustering method. Modify tables as necessary so that they all have the following format:

1. Rows are OTUs or SVs
2. The first column is the OTU/SV identifier, and the remaining columns are samples' abundances
3. There are no comment lines (just a header with column names)

After modifying the format of each table, write it back to a tab-separated '.txt' file.
```{r load inference results}

dataset_tables <- list()

for (d in datasets){
  resdir <- file.path(result_path, d)
  method_tables <- list()
  
  # load UCLUST table
  uclust_otu_file <- file.path(resdir, "uclust/otu_table.txt")
  uclust_seq_file <- file.path(resdir, "uclust/rep_set/pooled_nochim_rep_set.fasta")
  method_tables$uclust <- load_uclust(uclust_otu_file, uclust_seq_file)
  sample_name <- colnames(method_tables$uclust)[c(-1, -length(colnames(method_tables$uclust)))]
  method_tables$uclust <- remove_rare(method_tables$uclust, sample_name, min_abund = 2)  # remove singletons from all samples
  method_tables$uclust
  write_tsv(method_tables$uclust %>% select(-sequence), file.path(resdir, "uclust/uclust_table.txt"))
  
  # load UPARSE table
  uparse_otu_file <- file.path(resdir, "uparse/otu_table.txt")
  uparse_seq_file <- file.path(resdir, "uparse/otus.fa")
  method_tables$uparse <- load_uparse(uparse_otu_file, uparse_seq_file, sample_names = sample_name)
  method_tables$uparse <- remove_rare(method_tables$uparse, sample_name, min_abund = 2)  # remove singletons
  method_tables$uparse
  write_tsv(method_tables$uparse %>% select(-sequence), file.path(resdir, "uparse/uparse_table.txt"))
  
  # load MED table
  med_otu_file <- file.path(resdir, "med/MATRIX-COUNT.txt")
  med_seq_file <- file.path(resdir, "med/NODE-REPRESENTATIVES.fasta")
  med_chimera_file <- file.path(resdir, "med/uchime/NODE-chimeras.fasta")
  method_tables$med <- load_med(med_otu_file, med_seq_file, med_chimera_file)
  method_tables$med <- remove_rare(method_tables$med, sample_name, min_abund = 2)  # remove singletons
  method_tables$med
  write_tsv(method_tables$med %>% select(-sequence), file.path(resdir, "med/med_table.txt"))

  # load UNOISE table
  unoise_otu_file <- file.path(resdir, "unoise/zotu_table.txt")
  unoise_seq_file <- file.path(resdir, "unoise/zotus.fa")
  method_tables$unoise <- load_unoise(unoise_otu_file, unoise_seq_file, sample_names = sample_name)
  method_tables$unoise <- remove_rare(method_tables$unoise, sample_name, min_abund = 2)  # remove singletons
  method_tables$unoise
  write_tsv(method_tables$unoise %>% select(-sequence), file.path(resdir, "unoise/unoise_table.txt"))
  
  # load Deblur table
  deblur_otu_file <- file.path(resdir, "deblur/all.txt")
  method_tables$deblur <- load_deblur(deblur_otu_file)
  method_tables$deblur <- remove_rare(method_tables$deblur, sample_name, min_abund = 2)  # remove singletons
  method_tables$deblur
  write_tsv(method_tables$deblur %>% select(-sequence), file.path(resdir, "deblur/deblur_table.txt"))
  
  # load DADA2 table
  dada2_otu_file <- file.path(resdir, "dada2/sv_table.no_chim.txt")
  method_tables$dada2 <- load_dada2(dada2_otu_file, sample_names = sample_name)
  method_tables$dada2 <- remove_rare(method_tables$dada2, sample_name, min_abund = 2)  # remove singletons
  method_tables$dada2
  write_tsv(method_tables$dada2 %>% select(-sequence), file.path(resdir, "dada2/dada2_table.txt"))

  all_table <- merge_tables(method_tables, sample_name, collapse = TRUE)
  dataset_tables[[d]] <- all_table
}

```


Read in reference sequences, and compute Levenshtein ("Hamming") distances between all inferred sequences and all reference sequences.
```{r load reference fasta files}

zymo_fastas <- "Zymo_16S_references_v2.fasta"
kozich_fasta <- "kozich_HMP_MOCK.fasta"
schirmer_fasta <- "schirmer_BalancedRefSeqs.fasta"
damore_fasta <- "schirmer_BalancedRefSeqs.fasta"
ref_fastas <- list("zymo" = zymo_fastas, "kozich" = kozich_fasta, "schirmer" = schirmer_fasta, "damore" = damore_fasta)

ref_seqs <- lapply(ref_fastas, function(f){
  fp <- file.path(ref_path, f)
  seqs <- as.character(readDNAStringSet(fp))
  return(seqs)
})

names(ref_seqs$zymo) <- str_replace(names(ref_seqs$zymo), "_16S.*", "")
names(ref_seqs$kozich) <- str_replace(names(ref_seqs$kozich), "\\.\\d+$", "")

ref_dist_tables <- mapply(compute_ref_dist, dataset_tables, ref_seqs, SIMPLIFY = FALSE)
strain_dist_tables <- mapply(collapse_group_dist, ref_dist_tables, lapply(ref_seqs, names), SIMPLIFY = FALSE)

```


Check to see if any reference strains are completely absent from the dataset.
```{r missing strains}

# Start with just those reference strains not detected by any method
undetected <- lapply(strain_dist_tables, function(st){
  missing <- apply(st, 2, min) > 0
  missing <- names(missing[missing])
  return(missing)
})

# Get the corresponding reference sequences
undetected_refs <- mapply(function(refs, und){
  refs <- refs[names(refs) %in% und]
}, ref_seqs, undetected)

# Read in the (cleaned) fastq data
zymo_data <- "zymo_neat/filtered/pooled_filtered.fastq"
kozich_data <- "kozich_130403/filtered/pooled_filtered.fastq"
schirmer_data <- "schirmer_balanced/filtered/pooled_filtered.fastq"
damore_data <- "damore_uneven/filtered/pooled_filtered.fastq"
data_fastqs <- list("zymo" = zymo_data, "kozich" = kozich_data, "schirmer" = schirmer_data, "damore" = damore_data)

data_seqs <- lapply(data_fastqs, function(f){
  fp <- file.path(data_path, f)
  seqs <- derepFastq(fp, verbose = TRUE)
  seqs <- names(seqs$uniques)
  return(seqs)
})

missing_strains <- mapply(get_missing_strains, data_seqs, undetected_refs)

# Remove missing references from the reference set and ref_dist_tables
ref_seqs <- mapply(function(refs, missing){
  return(refs[!names(refs) %in% missing])
}, ref_seqs, missing_strains)

ref_dist_tables <- mapply(function(rdist, missing){
  return(rdist[, !colnames(rdist) %in% missing])
}, ref_dist_tables, missing_strains)

```


Annotate sequences in sequence tables as "Reference" if they exactly match a reference sequence, or as "Ref Noisy" if they are within a specified Levenshtein distance range from a reference sequence. 
```{r annotate reference and ref_noisy}

max_dist <- 10
dataset_tables <- mapply(annotate_ref, dataset_tables, ref_dist_tables, 
                         MoreArgs = list(sample_names = methods, max_dist = max_dist),
                         SIMPLIFY = FALSE)

```


Next, write each method's inferred (singleton-free) sequences to a fasta file, for BLASTing against nt.
```{r write fastas}

mapply(write_fasta, dataset_tables, file.path(analysis_path, "blast", paste(datasets, "seqs.fasta", sep = "_")))

# BLAST sequences against the 'nt' database using the following command:
# blastn -query all_seqs.fasta -db nt -out all_blast.txt \
#        -task megablast -max_target_seqs 10 \
#        -outfmt "7 qseqid qlen sseqid slen sskingdoms ssciname pident length nident mismatch gapopen gaps qstart qend sstart send evalue bitscore" \
#        -remote

```


Read in the results of the BLAST searches. 
```{r load and annotate blast tables}

blast_files <- file.path(analysis_path, "blast", paste(datasets, "blast.txt", sep = "_"))
blast_tables <- lapply(blast_files, load_blast)
names(blast_tables) <- datasets
blast_tables <- lapply(blast_tables, annotate_blast_table)

```


Annotate sequences with distance to the nearest NT sequence, and label them as 'Contaminant' if they exactly match an NT sequence, using the results of the BLAST search.
```{r annotate contaminant}

dataset_tables <- mapply(annotate_nt_dist, dataset_tables, blast_tables, SIMPLIFY = FALSE)

dataset_tables <- mapply(annotate_contam, dataset_tables, blast_tables, 
                        MoreArgs = list(sample_names = methods, max_dist = 10), 
                        SIMPLIFY = FALSE)

```


Compute distance from sequences to Contaminant sequences.
```{r compute distance to contaminant}

contam_seqs <- lapply(dataset_tables, function(dt) return(dt %>% filter(contaminant) %>% .[["sequence"]]))
contam_dist_tables <- mapply(compute_ref_dist, dataset_tables, contam_seqs, SIMPLIFY = FALSE)

```


Annotate Contam_Noisy, Other, and Consensus, then add a 'Class' factor column with the sequence classification of each sequence, with the five main classification levels.
```{r annotate contam noisy and contam mutant}

dataset_tables <- mapply(annotate_contam_like, dataset_tables, contam_dist_tables, blast_tables,
                        MoreArgs = list(sample_names = methods, noisy_dist = 10, mut_dist = 1),
                        SIMPLIFY = FALSE)

dataset_tables <- lapply(dataset_tables, annotate_class)

```


Compute distances between `contam_noisy` and `contaminant` sequences. This should give a sense of how many `contam_noisy` sequences may have been generated through sequencer errors from `contaminant` template sequences. Also, compute distances between `contaminant` sequences. This will give an idea of whether or not OTU methods underestimate the number of `contaminant` sequences by lumping them together, if they are within the 97% similarity threshold of each other.
```{r contam_noisy distances}

# noisy_to_contam <- lapply(dataset_tables, compute_inter_dist, class1 = "contam_noisy", class2 = "contaminant")
# dataset_tables <- mapply(annotate_inter_dist, dataset_tables, noisy_to_contam, 
#                         MoreArgs = list(column_name = "dist_to_contam"), SIMPLIFY = FALSE)

# table minimum distances of contam_noisy sequences to contaminant sequences
# for (d in datasets){
#   cat(d, "distances of contam_noisy to contaminant sequences:\n")
#   if (length(noisy_to_contam[[d]]) > 0){
#     print(noisy_to_contam[[d]] %>% apply(., 1, min) %>% table())
#   } else cat("There were no Contam Noisy sequences for the", d, "dataset.\n")
# }


# contam_to_contam <- lapply(dataset_tables, compute_inter_dist, class1 = "contaminant", class2 = "contaminant")
# dataset_tables <- mapply(annotate_inter_dist, dataset_tables, contam_to_contam, 
#                         MoreArgs = list(column_name = "dist_to_contam"), SIMPLIFY = FALSE)

# table minimum distances of contam_noisy sequences to contaminant sequences
# for (d in datasets){
#   cat(d, "distances of contaminant to contaminant sequences:\n")
#   print(contam_to_contam[[d]] %>% apply(., 1, function(d) min(d[d > 0])) %>% table())
#   cat("\n")
# }

```


Compute a table of sequence summaries for each method.
```{r compute sequence summaries}


dataset_summary <- mapply(summarize_seqs, dataset_tables, ref_dist_tables, ref_seqs,
                              MoreArgs = list(sample_names = methods, max_dist = max_dist),
                              SIMPLIFY = FALSE)
dataset_summary <- lapply(dataset_summary, function(ds){
  colnames(ds)[1] <- "method"
  return(ds)
})
dataset_summary <- lapply(dataset_summary, sanity_check_summary)
lapply(dataset_summary, function(ds) sum(ds$total != ds$check_sum))


dataset_summary_short <- lapply(dataset_summary, function(ds){
  dss <- ds %>% select(method, total, reference, ref_noisy, contaminant, contam_noisy, other)
  dss$method <- c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2")
  return(dss)
})

write_tables(dataset_summary, file.path(analysis_path, "summary", "dataset_summary_tables.tsv"))
write_tables(dataset_summary_short, file.path(analysis_path, "summary", "dataset_summary_short_tables.tsv"), compact = FALSE)


```


Assign taxonomy to the sequences.
```{r assign taxonomy}

# genus_db <- "silva_nr_v132_train_set.fa.gz"
# species_db <- "silva_species_assignment_v132.fa.gz"
# taxa <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
# 
# dataset_tables <- lapply(dataset_tables, assign_taxonomy, 
#                          genus_db_path = file.path(ref_path, genus_db), 
#                          species_db_path = file.path(ref_path, species_db))

```


Now, compute summary tables at the taxonomy level
```{r compute taxonomy summaries}

# dataset_taxa_summary <- lapply(dataset_tables, summarize_taxa, sample_names = methods, group = method)
# dataset_taxa_summary <- lapply(dataset_taxa_summary, function(dts){
#   dts$method <- factor(dts$method,
#                        levels = c("uclust", "uparse", "med", "unoise", "deblur", "dada2"),
#                        labels = c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2"))
#   dts <- dts %>% arrange(method)
#   return(dts)
# })
# write_tables(dataset_taxa_summary, file.path(analysis_path, "dataset_taxa_summary_tables.tsv"), compact = TRUE)

```


From the summary table, compute recall and precision for the sequences inferred by each method.
```{r method sequence summary stats}

dataset_seq_stats <- lapply(dataset_summary, compute_pr_seqs)
dataset_read_stats <- lapply(dataset_tables, compute_pr_reads, methods)
dataset_ref_perc <- lapply(dataset_tables, compute_ref_perc, sample_names = methods)
dataset_ref_perc <- do.call("cbind", dataset_ref_perc) %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as.tibble()

dataset_seq_stats <- lapply(dataset_seq_stats, function(dss){
  dss$method <- c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2")
  return(dss)
})

write_tables(dataset_seq_stats, file.path(analysis_path, "summary", "dataset_seq_stats.tsv"))
write_tables(dataset_read_stats, file.path(analysis_path, "summary", "dataset_read_stats.tsv"))
write_tsv(dataset_ref_perc, file.path(analysis_path, "summary", "dataset_ref_perc.tsv"))


```


Add a column that assigns a color code to each sequence
```{r add color column}

library(RColorBrewer)

ref_colors <- brewer.pal(8, "Dark2") %>% rev()
ref_colors <- c(brewer.pal(8, "Dark2") %>% rev(), brewer.pal(9, "Set1"), brewer.pal(8, "Accent") %>% rev(),
                brewer.pal(8, "Set2"), brewer.pal(12, "Paired"), brewer.pal(12, "Set3"), brewer.pal(11, "Spectral"))
ref_colors <- c(ref_colors, brewer.pal(3, "Set1"))
max_contams <- max(sapply(dataset_tables, function(mt) nrow(mt[!mt$reference,])))
other_colors <- sample(grey.colors(5, start = 0.5, end = 0.9), size = max_contams, replace = TRUE)
other_colors <- rep("#000000", times = max(sapply(dataset_tables, function(dt) nrow(dt[!dt$reference,]))))

dataset_tables <- lapply(dataset_tables, function(dt){
  dt$color <- character(length = nrow(dt))
  dt$color[dt$reference] <- ref_colors[1:nrow(dt[dt$reference,])]
  dt$color[!dt$reference] <- other_colors[1:nrow(dt[!dt$reference,])]
  return(dt)
})

```


Reformat annotated sequence tables for plotting with `ggplot`.
```{r prep for ggplot}

# Combine all method tables into one large table, then gather the sample columns into a single column
dataset_tables_gg <- lapply(dataset_tables, gather_samples, methods, method)
all_table_gg <- do.call("rbind", dataset_tables_gg) %>% mutate(dataset = rep(datasets, times = sapply(dataset_tables_gg, nrow)))
all_table_gg$dataset <- factor(all_table_gg$dataset, levels = c("zymo", "kozich", "schirmer", "damore"),
                              labels = c("Zymo", "Kozich", "Schirmer", "D'Amore"))
all_table_gg$method <- factor(all_table_gg$method, levels = c("uclust", "uparse", "med", "unoise", "deblur", "dada2"),
                              labels = c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2"))

all_table_gg <- annotate_norms(all_table_gg, group = c("dataset", "method"))

# add some grouping variables for plotting convenience
all_table_gg <- all_table_gg %>% mutate(algorithm = method)
levels(all_table_gg$algorithm) <- list(OTU = c("UCLUST", "UPARSE"), entropy = "MED", EM = c("UNOISE", "Deblur", "DADA2"))

# Grab an arbitrary zero-count, non-Reference row from Zymo for Deblur, to use as a filler
# This will be used for the SNR boxplot, since Deblur had no non-Reference sequences for Zymo
deblur_zero <- all_table_gg %>% filter(method == "Deblur", dataset == "Zymo", count == 0, class != "reference") %>% .[1, ]
deblur_zero$count <- 1

# Remove zero-count rows
all_table_gg <- all_table_gg %>% filter(count > 0)

```


##Create various plots to visualize results

Plot the sequence-level composition of samples by sequence class.
```{r class seqs composition}

library(ggpubr)

# library(devtools)
# unloadNamespace("ggbeeswarm")
# unloadNamespace("ggridges")
# unloadNamespace("dada2")
# unloadNamespace("tidyverse")
# unloadNamespace("ggplot2")
# unloadNamespace("ggpubr")
# unloadNamespace("cowplot")
# dev_mode(TRUE)
# # install_github("hadley/ggplot2")
# library(ggplot2)
# packageVersion("ggplot2")
# dev_mode(F)
# library(tidyverse)
# 
theme_set(theme_bw())
classes = levels(all_table_gg$class)
class_colors <- brewer.pal(12, "Paired")[c(2, 1, 8, 7, 10)]
# 
# class_seqs_comp <- ggplot(all_table_gg %>% filter(count > 0), aes(x = method)) +
#   geom_bar(aes(fill = class), width = 3, position = position_dodge(preserve = "single")) +
#   scale_fill_manual(name = "Sequence class", values = class_colors) +
#   labs(x = "method", y = "number of sequences") +
#   facet_wrap(~dataset)
# class_seqs_comp
# 
# unloadNamespace("tidyverse")
# unloadNamespace("ggplot2")
# library(ggplot2)
# packageVersion("ggplot2")
# library(tidyverse)

class_seqs_comp <- ggplot(all_table_gg, aes(x = method)) +
  geom_bar(aes(fill = fct_rev(class)), width = 0.65) +
  facet_wrap(~dataset) +
  scale_fill_manual(name = "ASV/OTU class   ", values = rev(class_colors), labels = paste(rev(classes), "   ")) +
  labs(x = "", y = "number of ASVs/OTUs") +
  big_labels(title = 72, text = 60, legend.position = "top") +
  theme(panel.spacing = unit(3, "lines"), 
        legend.margin = margin(b = 90)) +
  guides(fill = guide_legend(reverse = TRUE, keywidth = 4.5, keyheight = 4.5))
class_seqs_comp


class_reads_comp <- ggplot(all_table_gg, aes(x = method)) +
  geom_col(aes(y = count, fill = fct_rev(class)), position = position_fill(), width = 0.65) +
  scale_fill_manual(name = "ASV/OTU class", values = rev(class_colors),
                    guide = guide_legend(reverse = FALSE)) +
  facet_wrap(~dataset) +
  # labs(title = "Read-level composition by sequence class",
  labs(x = "method", y = "fraction of sample") +
  big_labels(title = 72, text = 60, legend.position = "top") +
  theme(panel.spacing = unit(3, "lines"), 
        legend.margin = margin(b = 90)) +
  guides(fill = guide_legend(reverse = TRUE, keywidth = 4.5, keyheight = 4.5))
  # guides(fill = FALSE)
class_reads_comp

ggarrange(class_seqs_comp, class_reads_comp, labels = c("A", "B"), vjust = c(1.5, 0.5), nrow = 2, 
          font.label = list(size = 96), heights = c(5, 4))

# class_counts <- all_table_gg %>% 
#   filter(count > 0) %>%
#   group_by(dataset, method, class) %>% 
#   summarize(count = n()) %>% 
#   complete(class, fill = list(count = 0))

# class_seqs_comp <- ggplot(class_counts, aes(x = method)) +
#   geom_col(aes(y = count, fill = fct_rev(class)), width = 0.6, position = position_dodge()) +
#   scale_fill_manual(name = "Sequence class   ", values = rev(class_colors), labels = paste(rev(classes), "   ")) +
#   labs(x = "method", y = "number of sequences") +
#   facet_wrap(~dataset) +
#   big_labels(title = 72, text = 60, legend.position = "top") +
#   theme(panel.spacing = unit(3, "lines"), 
#         legend.margin = margin(b = 90)) +
#   guides(fill = guide_legend(keywidth = 4.5, keyheight = 4.5))
# class_seqs_comp

```



Plot the read-level composition of samples by sequence class. 
```{r class reads composition}

# class_colors <- brewer.pal(9, "Set1")[c(9, 5, 4, 1, 2)]
# class_colors <- brewer.pal(9, "Set1")[c(2, 3, 1, 4, 5)]
# class_colors <- brewer.pal(8, "Dark2")[c(8, 2, 3, 4, 1)]
# class_colors <- brewer.pal(8, "Accent")[c(8, 7, 1, 6, 5)]
# class_colors <- brewer.pal(12, "Paired")[c(2, 1, 6, 5, 10)]
# 
# 
# class_reads_comp_dataset <- ggplot(all_table_gg %>% filter(count > 0), aes(x = method)) +
#   geom_col(aes(y = count, fill = fct_rev(class)), position = position_fill(), width = 0.5) +
#   scale_fill_manual(name = "Sequence class", values = rev(class_colors), 
#                     guide = guide_legend(reverse = FALSE)) +
#   facet_wrap(~dataset) +
#   labs(title = "Read-level composition by sequence class",
#        subtitle = "", y = "relative abundance") +
#   big_labels(angle = 30, hjust = 0.7, vjust = 0.7, legend.position = "right")
# class_reads_comp_dataset

```


Create plot to show sample composition, colored by reference vs. non-reference.
```{r composition plot}

mname = "UCLUST"

set.seed(444)
# data_table <- method_tables[[tolower(mname)]]
# data_gg <- all_table_gg %>% filter(method == mname)


# ref_colors <- brewer.pal(sum(data_table$reference), "Dark2") %>% rev()
# ref_colors <- brewer.pal(8, "Dark2") %>% rev()
# ref_colors <- c(ref_colors, brewer.pal(3, "Set1")[1])
# other_colors <- sample(grey.colors(5, start = 0.5, end = 0.9), sum(!data_table$reference), replace = TRUE)

# data_table$color <- rep(NA, nrow(data_table))
# data_table$color[data_table$reference] <- ref_colors
# data_table$color[!data_table$reference] <- other_colors

# color_gg <- data_table %>% select(id, sample_names, color) %>% gather("sample", "count", sample_names)
# data_gg <- inner_join(data_gg, color_gg)
# data_gg <- data_table %>% gather("sample", "count", sample_names)

comp_bar <- ggplot(all_table_gg %>% filter(method == mname), aes(x = sample, y = count)) +
  geom_col(aes(fill = color, group = reference), width = 0.5, position = position_fill()) +  #alpha = ifelse(reference, "bright", "dull")), 
  scale_fill_identity(name = "Organism", guide = "legend", 
                      labels = c(paste("Ref", 1:8), rep("non-Ref", length(other_colors))), 
                      breaks = c(ref_colors, other_colors)) +
  scale_x_discrete(labels = dilution_labels) +
  labs(title = "Relative composition of dilution series samples", 
       subtitle = paste("Dilution series processed with", mname), x = "sample dilution", y = "fraction of sample") +
  theme(legend.position = "right", legend.title = element_text(size = 12), 
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        plot.title = element_text(size = 20))
comp_bar
comp_bar + coord_polar(theta = "y", start = 0)

comp_bar_facet <- ggplot(all_table_gg, aes(x = method, y = count)) +
  geom_col(aes(fill = color, group = reference), width = 0.5, position = position_fill()) +
  scale_fill_identity(name = "ASV/OTU", guide = "legend",
                      labels = c(paste("Ref", 1:58), rep("non-Ref", length(other_colors))),
                      breaks = c(ref_colors[1:58], other_colors)) +
  scale_x_discrete(labels = methods) +
  facet_wrap(~dataset) +
  # labs(title = "Composition of high biomass samples", 
       labs(x = "method", y = "fraction of sample") +
  theme(legend.position = "right",
        legend.title = element_text(face = "bold", size = 12), 
        axis.text = element_text(size = 12),
        axis.title = element_text(face = "bold", size = 16),
        plot.title = element_text(size = 20),
        axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
        strip.text = element_text(size = 16),
        panel.spacing = unit(1, "lines"))
comp_bar_facet

```


Plot alpha diversity.
```{r alpha diversity}

library("phyloseq")

# extract sequence matrices of read counts only from the method sequence tibbles
seq_mats <- lapply(dataset_tables, function(mt){
  seq_mat <- as.matrix(mt[, methods])
  rownames(seq_mat) <- mt$id
  return(seq_mat)
})


# construct phyloseq OTU and taxonomy tables
otu_tabs <- lapply(seq_mats, otu_table, taxa_are_rows = TRUE)

# construct phyloseq objects
physeqs <- lapply(otu_tabs, phyloseq)

# compute alpha diversity
alpha_indices <- c("Chao1", "Shannon", "Simpson", "InvSimpson", "Fisher")
alpha_div <- lapply(physeqs, estimate_richness, measures = alpha_indices)

# combine alpha diversity tables into one table for plotting
alpha_div <- lapply(alpha_div, rownames_to_column, var = "method")
alpha_div <- lapply(alpha_div, as.tibble)
alpha_div_gg <- lapply(alpha_div, gather, key = "alpha_index", value = "diversity", alpha_indices)
alpha_div_gg <- do.call("rbind", alpha_div_gg) %>% 
  mutate(dataset = rep(datasets, each = length(methods) * length(alpha_indices)))

alpha_colors <- brewer.pal(12, "Paired")[c(2, 3, 9, 8)]

alpha_dots <- ggplot(alpha_div_gg %>% filter(alpha_index %in% c("Shannon", "Simpson", "InvSimpson", "Fisher")), 
                     aes(x = method, group = dataset)) +
  geom_point(aes(y = diversity, color = dataset), size = 2.5) +
  scale_color_manual(values = alpha_colors) +
  facet_wrap(~alpha_index, nrow = 1, scales = "free_y") +
  big_labels(legend.position = "top")
alpha_dots


```


Create stacked histograms showing numbers of sequences vs. read count for reference and non-reference sequences.
```{r read count histograms}

# theme_set(theme_bw())
# 
# library(ggridges)
# 
# min_count = 0
# dname = "Kozich"
# 
# read_hist <- ggplot(all_table_gg %>% filter(dataset == dname, count > min_count), aes(x = count, y = reference)) +
#   geom_density_ridges(aes(fill = reference, height = ..density..), scale = 0.9, stat = "binline", bins = 120, panel_scaling = FALSE) +
#   geom_vline(data = all_table_gg %>% filter(reference) %>% group_by(method) %>% summarise(min = min(count)), aes(xintercept = min)) +
#   labs(title = "Histograms of abundances of Reference vs. non-Reference sequences", x = "abundance", y = "Sequence type") +
#   facet_wrap(~method) +
#   scale_fill_brewer(name = "Sequence type", palette = "Set1", labels = c("non-Reference", "Reference")) +
#   big_labels()
# read_hist

```


Create plots to show relative strength of signal (reference sequence matches) to noise (non-reference sequences).
```{r snr plots}

# library(ggbeeswarm)
library(scales)
# library(cowplot)
# library(gridExtra)

# dname = "Schirmer"
# 
# # ref vs. non-ref boxplots, log10 raw counts
# snr_box <- ggplot(data = all_table_gg %>% filter(dataset == dname, count > 0), 
#                   aes(x = method, y = log10_count)) +
#   geom_boxplot(aes(fill = factor(reference, levels = c(TRUE, FALSE)))) +
#   labs(title = "Distribution of log10(read counts) for reference and non-reference sequences",
#           subtitle = "Dilution series 'neat' sample", x = "sample dilution", y = "log10(read counts)") +
#   scale_fill_manual(name = "Sequence type", labels = c("Reference", "non-Reference"), values = c("blue", "orange")) +
#   big_labels()
# snr_box

snr_box_facet <- ggplot(data = rbind(all_table_gg, deblur_zero)) +
  geom_boxplot(aes(x = method, y = count, fill = factor(reference, levels = c(TRUE, FALSE))),
               outlier.size = 3, outlier.stroke = 1, size = 1.5) +
  labs(x = "method", y = expression(bold("log"[10]*" abundance"))) +
  scale_fill_manual(name = "ASV/OTU class   ", 
                    labels = paste(c(" Reference", " non-Reference"), "   "), 
                    values = c("blue", "orange")) +
  scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
        labels = trans_format("log10", math_format(10^.x))) +
  facet_wrap(~dataset) +
  big_labels(title = 72, text = 60, legend.position = "top") +
  theme(panel.spacing = unit(3, "lines"),
        legend.margin = margin(b = 90)) +
  guides(fill = guide_legend(keywidth = 6, keyheight = 9))
snr_box_facet


# ref vs. non-ref beeswarm plots, log10 raw counts
# snr_bees <- ggplot(data = all_table_gg %>% filter(dataset == dname, count > 0), 
#                    aes(x = method, y = log10_count)) +
#   geom_beeswarm(aes(color = factor(reference, levels = c(T, F))), priority = "ascending", cex = 0.15, dodge.width = 0.5) +
#   # geom_quasirandom(aes(color = factor(reference, levels = c(T, F))), dodge.width = 0.75, bandwidth = 0.5) +
#   labs(title = "Density distribution of log10(read counts) for reference and non-reference sequences",
#           subtitle = "Dilution series 'neat' sample", x = "method", y = "log10(read counts)") +
#   scale_color_manual(name = "sequence identity", labels = c("reference", "non-reference"), values = c("blue", "orange")) +
#   big_labels()
# snr_bees


# snr_bees_facet <- ggplot(data = all_table_gg %>% filter(count > 0)) +
#   geom_beeswarm(aes(x = method, y = log10_count, color = factor(reference, levels = c(T, F))), size = 2, cex = 0.05, dodge.width = 0.3) +
#   # labs(title = "Density distribution of log10(read counts) for reference and non-reference sequences",
#   #         subtitle = paste("Dilution series processed with various methods"), x = "sample dilution", y = "log10(read counts)") +
#   labs(x = "sample dilution", y = "log10(read count)") +
#   scale_color_manual(name = "sequence identity", labels = c("Reference", "non-Reference"), values = c("blue", "orange")) +
#   scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
#         labels = trans_format("log10", math_format(10^.x))) +
#   facet_wrap(~dataset, nrow = 2) +
#   big_labels(legend.position = "top")
# snr_bees_facet + guides(color = guide_legend(override.aes = list(size = 3)))


# box_legend <- get_legend(snr_box_facet)
# box_legend <- plot_grid(box_legend, ncol = 1)
# box_legend
# 
# grid.arrange(snr_box_facet, box_legend, ncol = 1)

# ref vs. non-ref violin plots, log10 raw counts
# snr_violin <- ggplot(data = all_table_gg %>% filter(dataset == dname, count > 0), 
#                      aes(x = method, y = log10_count)) +
#   geom_violin(aes(fill = factor(reference, levels = c(T, F)))) +
#   labs(title = "Density distribution of log10(read counts) for reference and non-reference sequences",
#           subtitle = "Dilution series 'neat' sample", x = "sample dilution", y = "log10(read counts)") +
#   scale_fill_manual(name = "sequence identity", labels = c("reference", "non-reference"), values = c("blue", "orange")) +
#   big_labels()
# snr_violin

# snr_jitter <- ggplot(data = all_table_gg %>% filter(dataset == dname, count > 0), 
#                      aes(x = method, y = log10_count)) +
#   geom_jitter(aes(color = factor(reference, levels = c(T, F))), width = 0.1, size = 2) +
#   labs(title = "Density distribution of log10(read counts) for reference and non-reference sequences",
#           subtitle = "Dilution series 'neat' sample", x = "sample dilution", y = "log10(read counts)") +
#   scale_color_manual(name = "sequence identity", labels = c("reference", "non-reference"), values = c("blue", "orange")) +
#   big_labels()
# snr_jitter


# average (mean(log10)) of ref vs. non-ref signal
# mean_snr_line <- ggplot(sample_sn_gg, aes(x = sample, y = mean_log10)) +
#   geom_point(data = subset(sample_sn_gg, reference), aes(color = method, group = method)) +
#   geom_line(data = subset(sample_sn_gg, reference), aes(color = method, group = method)) + 
#   geom_point(data = subset(sample_sn_gg, !reference), aes(color = method, group = method)) +
#   geom_line(data = subset(sample_sn_gg, !reference), aes(color = method, group =method))
# mean_snr_line
# 
# mean_snr_line <- ggplot(sample_summary, aes(x = sample, y = mean_snr)) +
#   geom_point(aes(color = method, group = method)) +
#   geom_line(aes(color = method, group = method)) +
#   scale_color_brewer(palette = "Dark2")
# mean_snr_line

```

Create plot to show sample composition, colored by reference vs. non-reference.
```{r composition plot}

# library(RColorBrewer)
# 
# set.seed(444)
# 
# dname = "Zymo"
# 
# ref_colors <- brewer.pal(sum(dataset_tables$zymo$reference), "Dark2")
# other_colors <- sample(grey.colors(5, start = 0.5, end = 0.9), sum(!dataset_tables$zymo$reference), replace = TRUE)
# 
# dataset_tables$zymo$color <- rep(NA, nrow(dataset_tables$zymo))
# dataset_tables$zymo$color[dataset_tables$zymo$reference] <- ref_colors
# dataset_tables$zymo$color[!dataset_tables$zymo$reference] <- other_colors
# 
# 
# color_gg <- dataset_tables$zymo %>% select(id, methods, color) %>% gather("method", "count", methods)
# legend_color <- c(bright = ref_colors[2], dull = other_colors[2])
# all_table_gg <- inner_join(all_table_gg, color_gg)
# 
# comp_bar <- ggplot(all_table_gg, aes(x = method, y = count)) +
#   geom_col(aes(fill = color, group = reference, alpha = ifelse(reference, "bright", "dull")), width = 0.5, 
#            position = position_fill()) +
#   scale_fill_identity(guide = FALSE) +
#   scale_alpha_manual(name = "Sequence type", 
#                      labels = c("mock community", "other"),
#                      values = c(bright = 1, dull = 1), 
#                      guide = guide_legend(override.aes = list(fill = c(ref_colors[4], "#AEAEAE")), 
#                                           keywidth = NULL, keyheight = NULL)) +
#   labs(title = "Relative composition sample processed by different methods", 
#        subtitle = "Zymo 'neat' sample",
#        x = "sample dilution", y = "fraction of sample") +
#   theme(legend.position = "right", legend.title = element_text(size = 12), 
#         axis.text = element_text(size = 12),
#         axis.title = element_text(size = 16),
#         plot.title = element_text(size = 20))
# comp_bar
#comp_bar + coord_polar(theta = "y", start = 0)

```


Plot bar plots of relative abundance of reference sequences, facetted by method.
```{r reference seq bar plots}

# dname = "Zymo"
# 
# ref_bars <- ggplot(data = all_table_gg %>% filter(dataset == dname, reference)) +
#   geom_col(aes(x = id, y = rel_count, fill = id), width = 0.8) +
#   facet_wrap(~method) +
#   scale_fill_brewer(palette = "Dark2") +
#   big_labels()
# ref_bars

# ref_joy <- ggplot(data = all_table_gg %>% filter(reference)) +
#   geom_density_ridges(aes(x = id, y = method, fill = id, height = ..density..), stat = "density")
# ref_joy

```


Plot distribution boxplots of non-reference read counts for "contaminant" sequences vs non-contaminant ("other") sequences.
```{r snr contaminant vs other}

# dname = "Kozich"
# 
# snr_box_contam <- ggplot(all_table_gg %>% filter(dataset == dname, count > 0, !ref_like), 
#                          aes(x = method, y = log10_count)) +
#   geom_boxplot(aes(fill = factor(contaminant, levels = c(TRUE, FALSE)))) +
#   labs(title = "Distribution of log10(read counts) for non-reference contaminant and non-contaminant sequences",
#           subtitle = "Dilution series 'neat' sample processed with various methods", x = "sample dilution", y = "log10(read counts)") +
#   scale_fill_manual(name = "sequence identity", labels = c("contaminant", "non-contaminant"), values = c("blue", "orange")) +
#   big_labels()
# snr_box_contam

```


Plot bar plots of summary stats (number of "reference", "ref_noisy", "contaminant", "other").
```{r summary stat bar plots}

# summary_gg <- gather(summary_table, "statistic", "value", -method) %>% as.tibble
# 
# ref_mean <- all_table_gg %>% filter(reference) %>% .[["count"]] %>% mean()
# ref_noisy_mean <- all_table_gg %>% filter(ref_noisy) %>% .[["count"]] %>% mean()
# contam_mean <- all_table_gg %>% filter(contaminant) %>% .[["count"]] %>% mean()
# contam_noisy_mean <- all_table_gg %>% filter(contam_noisy) %>% .[["count"]] %>% mean()
# other_mean <- all_table_gg %>% filter(other) %>% .[["count"]] %>% mean()
# 
# summary_gg <- summary_gg %>% mutate(norm_mean = rep(1, nrow(summary_gg)))
# summary_gg[summary_gg$statistic == "reference", "norm_mean"] <- summary_gg %>% 
#   filter(statistic == "reference") %>% .[["value"]] * log10(ref_mean)
# summary_gg[summary_gg$statistic == "ref_noisy", "norm_mean"] <- summary_gg %>% 
#   filter(statistic == "ref_noisy") %>% .[["value"]] * log10(ref_noisy_mean)
# summary_gg[summary_gg$statistic == "contaminant", "norm_mean"] <- summary_gg %>% 
#   filter(statistic == "contaminant") %>% .[["value"]] * log10(contam_mean)
# summary_gg[summary_gg$statistic == "contam_noisy", "norm_mean"] <- summary_gg %>% 
#   filter(statistic == "contam_noisy") %>% .[["value"]] * log10(contam_noisy_mean)
# summary_gg[summary_gg$statistic == "other", "norm_mean"] <- summary_gg %>% 
#   filter(statistic == "other") %>% .[["value"]] * log10(other_mean)
# 
# bar_colors <- brewer.pal(9, "Set1")
# 
# summary_bars <- ggplot(data = summary_gg %>% filter(statistic %in% c("reference", "ref_noisy", "contaminant", "contam_noisy", "other")), 
#                        aes(x = statistic)) +
#   geom_col(aes(y = log2(1 + value), fill = statistic), position = position_dodge()) +
#   facet_wrap(~method) +
#   scale_fill_manual(values = ref_colors[c(1, 3, 4, 6, 8)])
#   #scale_fill_brewer(palette = "Set1")
# summary_bars
# 
# summary_pct_bars <- ggplot(data = summary_gg %>% filter(statistic %in% c("pct_ref", "pct_ref_noisy", "pct_contam", "pct_contam_noisy", "pct_other")), 
#                        aes(x = statistic)) +
#   geom_col(aes(y = value, fill = statistic), position = position_dodge()) +
#   facet_wrap(~method) +
#   scale_fill_manual(values = bar_colors[c(3, 4, 5, 2, 1)])
#   #scale_fill_brewer(palette = "Set1")
# summary_pct_bars

```
